---
title: "FastReg"
author: "Matt W. Wheeler, Shail Choksi, Deepak Mav, Adam Burkholder, Fred Wright, Alison A. Motsinger-Reif, John S. House"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    theme: united
vignette: >
  %\VignetteIndexEntry{FastReg}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

In genome-wide association studies (GWAS) and other genomic analyses, logistic regression is a central analysis tool for binary responses. 
For millions of SNPs that require millions of regressions, logistic regression analyses require significant computational resources, 
which are especially costly in terms of time, money, and carbon footprint when carried out in typical analysis pipelines. 
Software that maximizes analysis throughput can reduce a researcher's costs when analyzing data.

This software, FastReg, performs linear and logistic regression for arbitrary covariates (i.e., continuous, count, and binary covariates) 
at 15% faster speeds than highly specialized tools such as PLINK 2 on linux operating systems, and is 100x more efficient than serially 
using GLM in R.

# Pre-requisites and Installation
## Windows

- Install [**Rtools for Windows**](https://cran.r-project.org/bin/windows/Rtools/ "Rtools for Windows")

## MacOS

- Install [**macrtools**](https://github.com/coatless-mac/macrtools "macrtools")
- Start R and run the following:
```
install.packages("remotes")
remotes::install_github("coatless-mac/macrtools")
macrtools::xcode_cli_install()
macrtools::gfortran_install()
```
- This will attempt to install Xcode CLI and gfortran that are required to install FastReg

## Installing FastReg

```
# install dependencies
install.packages(c("Rcpp", "RcppArmadillo", "stats", "data.table", "BiocManager", "devtools", "BEDMatrix"))
Biocmanager::install("rhdf5")
library(devtools)
# install the package from github
install_github("SciomeLLC/FastReg")
library(FastReg)
```

# Dataset preparation

Let's look at an example GWAS dataset of 500 individuals x 5000 POIs stored as text files:

-   testdata.bin.pheno.txt - contains 500 individual ids and the binary response for each

    ```{r include=TRUE}
    # Binary phenotype data
    read.csv("../inst/extdata/testdata_500_by_1k.bin.pheno.txt", sep = "\t", nrows = 10)
    ```

-   testdata.covar.txt - contains the individual ids, and 3 covariates (age, sex and treatment)

    ```{r include=TRUE}
    # Covariate data
    read.csv("../inst/extdata/testdata_500_by_1k.covar.txt", sep = "\t", nrows = 10)
    ```

-   testdata.poi.txt - contains 500 individual ids and 5000 Predictors of Interest
<details>
  <summary>Click to see contents (large output)</summary>
    ```{r eval=TRUE, include=TRUE}
    # POI data
    read.csv("../inst/extdata/testdata_500_by_1k.poi.txt", sep = "\t", nrows = 2)
    ```
</details>

## Converting to HDF5
FastReg provides functionality to convert the POI text dataset into HDF5 dataset(s). 
Lets create a HDF5 dataset using the poi text data:

```{r include=TRUE, eval=FALSE}
library(FastReg)
TextToH5("testdata", "./FastRegData", delimiter="\t")
```

This will create n (where n is optimized according to your machine configuration) HDF5 datasets in the "FastRegData" directory.
# Regression 
Now we are ready to run fast regression on the dataset. 

## Logistic Regression
With the text file converted to HDF5 format, we are ready to run logistic regression on the full dataset.

```{r include=TRUE, eval=FALSE}
FastReg(
  regression.type="logistic",
  pheno.file="inst/extdata/testdata_500_by_1k.bin.pheno.txt",
  pheno.rowname.cols="ind",
  phenotype = "bin.resp",
  covar.rownames.cols="ind",
  covar.file="inst/extdata/testdata_500_by_1k.covar.txt",
  covariates=c("age", "sex", "treatment"),
  covariate.type = c("numeric", "categorical", "categorical"),
  covariate.standardize = c(TRUE, FALSE, FALSE),
  covariate.levels = c("", "F,M", "Placebo,Test"),
  covariate.ref.level = c("", "F", "Placebo"),
  POI.file.dir = "./FastRegData",
  POI.type = "dosage",
  POI.effect.type = "additive",
  output.dir = "FastRegResults"
)
```
The parameters for the covariates define the structure, data type, standardization, levels and reference levels. These parameters are all a list of the same size and need to be defined in the same order for each. The regression will try and spawn n processes and process the dataset in parallel. Two files will be created in the "FastRegResults" directory - one for the convergence and the other with the results of the logistic regression.

```{r include=TRUE, eval=FALSE}
read.csv("FastRegResults/Full_Convergence_stratum_1.tsv", delimiter="\t", nrows=10)
read.csv("FastRegResults/Full_Results_stratum_1.tsv", delimited="\t", nrows=10)
```

## Linear Regression

```{r include=TRUE, eval=FALSE}
FastReg::FastReg(
    phenotype="bin.resp", 
    max.iter = 6, 
    covariates = c("age","sex","treatment"),
    covariate.levels = ";F,M;Placebo,Test", # vectors / list of vectors 
    covariate.ref.level = ",F,Placebo", 
    covariate.type = "numeric,categorical,categorical",
    covariate.standardize = "1,0,0",
    covar.file= data(paste0(prefix, ".covar.txt"), package=FastReg),
    poi.file = data(paste0(prefix, ".poi.txt"), package=FastReg),
    pheno.file= data(paste0(prefix, ".bin.pheno.txt"), package=FastReg),
    regression.type="linear",
    output.dir = "."
)
```

## Optimization and OS specific settings

### OpenMP Threads

FastReg makes use of OpenMP threads and allows the user to set OpenMP threads per worker. The `max.openmp.threads` parameter can be passed to the FastReg function to set the number of OpenMP threads that will be used. Our benchmarking suggests that the default value of 1 openMP thread is usually optimal: 

```{r include=TRUE, eval=FALSE}
FastReg(
  regression.type="logistic",
  pheno.file="inst/extdata/testdata.bin.pheno.txt",
  pheno.rowname.cols="ind",
  phenotype = "bin.resp",
  covar.file="inst/extdata/testdata.covar.txt",
  covariates=c("age", "sex", "treatment"),
  covariate.type = c("numeric", "categorical", "categorical"),
  covariate.standardize = c(TRUE, FALSE, FALSE),
  covariate.levels = c("", "F,M", "Placebo,Test"),
  covariate.ref.level = c("", "F", "Placebo"),
  POI.file.dir = "./FastRegData"
  POI.type = "dosage",
  POI.effect.type = "additive",
  output.dir = "FastRegResults",
  max.openmp.threads = 4
)
```
NOTE: MacOS doesn't support OpenMP natively currently and this parameter will be defaulted to 1 on a MacOS system.

### Workers

FastReg will spawn min(Total POI files, # cores-1) of workers by default when running regressions. If you need granular control on how many workers are created, you can use the `max.workers` parameter to limit the workers.

```{r include=TRUE, eval=FALSE}
FastReg(
  regression.type="logistic",
  pheno.file="inst/extdata/testdata.bin.pheno.txt",
  pheno.rowname.cols="ind",
  phenotype = "bin.resp",
  covar.file="inst/extdata/testdata.covar.txt",
  covariates=c("age", "sex", "treatment"),
  covariate.type = c("numeric", "categorical", "categorical"),
  covariate.standardize = c(TRUE, FALSE, FALSE),
  covariate.levels = c("", "F,M", "Placebo,Test"),
  covariate.ref.level = c("", "F", "Placebo"),
  POI.file.dir = "./FastRegData"
  POI.type = "dosage",
  POI.effect.type = "additive",
  output.dir = "FastRegResults",
  max.workers = 4
)
```
NOTE: This parameter doesn't have any effect on Windows system as windows doesn't support fork().

### Windows optimization

Since Windows doesn't natively support fork(), we recommend using `max.openmp.threads` and setting this value to what you would have set for `max.workers`. Our testing suggests that this is similar in performance to `max.workers` on Linux systems. E.g. if the system has 10 cores: 

```{r include=TRUE, eval=FALSE}
FastReg(
  regression.type="logistic",
  pheno.file="inst/extdata/testdata.bin.pheno.txt",
  pheno.rowname.cols="ind",
  phenotype = "bin.resp",
  covar.file="inst/extdata/testdata.covar.txt",
  covariates=c("age", "sex", "treatment"),
  covariate.type = c("numeric", "categorical", "categorical"),
  covariate.standardize = c(TRUE, FALSE, FALSE),
  covariate.levels = c("", "F,M", "Placebo,Test"),
  covariate.ref.level = c("", "F", "Placebo"),
  POI.file.dir = "./FastRegData"
  POI.type = "dosage",
  POI.effect.type = "additive",
  output.dir = "FastRegResults",
  max.openmp.threads = 10
)
```

Windows users may get an error related to the HDF5 library: 
```
***HDF5 library version mismatched error***
```

You can set the following environment variable to safely disable the warning: 
```
Sys.setenv(HDF5_DISABLE_VERSION_CHECK='2')
```
FastReg will work as intended. 